# Filtering Data with SQL - Lab

## Introduction 

NASA wants to go to Mars! Before they build their rocket, NASA needs to track information about all of the planets in the Solar System. In this lab, you'll practice querying the database with various `SELECT` statements. This will include selecting different columns and implementing other SQL clauses like `WHERE` to return the data desired.

<img src="https://raw.githubusercontent.com/learn-co-curriculum/dsc-filtering-lab-v2-4/master/images/planets.png" alt="image of solar system" width="600">

## Objectives

You will practice the following:

* Retrieve a subset of records from a table using a `WHERE` clause
* Filter results using conditional operators such as `BETWEEN`, `IS NULL`, and `LIKE`
* Apply an aggregate function to the result of a filtered query

## Connecting to the Database

To get started, import `sqlite3` as well as `pandas` for conveniently displaying results. Then, connect to the SQLite database located at `planets.db`. 

In [1]:
# Your code here
import sqlite3
import pandas as pd

# Establish a connection to the SQLite database
conn = sqlite3.connect('planets.db')

# Write a SQL query
query = "SELECT * FROM planets"

# Execute the query and fetch the data into a pandas DataFrame
df = pd.read_sql_query(query, conn)

# Display the DataFrame
print(df)

   id     name       color  num_of_moons    mass  rings
0   1  Mercury        gray             0    0.55      0
1   2    Venus      yellow             0    0.82      0
2   3    Earth        blue             1    1.00      0
3   4     Mars         red             2    0.11      0
4   5  Jupiter      orange            68  317.90      0
5   6   Saturn       hazel            62   95.19      1
6   7   Uranus  light blue            27   14.54      1
7   8  Neptune   dark blue            14   17.15      1


## Database Schema

This database contains a single table, `planets`. This is the schema:

```
CREATE TABLE planets (
  id INTEGER PRIMARY KEY,
  name TEXT,
  color TEXT,
  num_of_moons INTEGER,
  mass REAL,
  rings BOOLEAN
);
```

The data looks something like this:

| id | name    | color      | num_of_moons | mass   | rings |
| -- | ------- | ---------- | ------------ | ------ | ----- |
| 1  | Mercury | gray       | 0            | 0.55   | FALSE |
| 2  | Venus   | yellow     | 0            | 0.82   | FALSE |
| 3  | Earth   | blue       | 1            | 1.00   | FALSE |
| 4  | Mars    | red        | 2            | 0.11   | FALSE |
| 5  | Jupiter | orange     | 67           | 317.90 | FALSE |
| 6  | Saturn  | hazel      | 62           | 95.19  | TRUE  |
| 7  | Uranus  | light blue | 27           | 14.54  | TRUE  |
| 8  | Neptune | dark blue  | 14           | 17.15  | TRUE  |

## SQL Queries

Write SQL queries for each of the statements below using the same pandas wrapping syntax from the previous lesson.

### 1. Select just the `name` and `color` of each planet

In [2]:
# Your code here
# 1. Select All Planets
query_all_planets = "SELECT * FROM planets"
df_all_planets = pd.read_sql_query(query_all_planets, conn)
print("All Planets:")
print(df_all_planets)

# 2. Select Planets with Rings
query_planets_with_rings = "SELECT * FROM planets WHERE rings = TRUE"
df_planets_with_rings = pd.read_sql_query(query_planets_with_rings, conn)
print("\nPlanets with Rings:")
print(df_planets_with_rings)

# 3. Select Planets with More Than 1 Moon
query_planets_more_than_1_moon = "SELECT * FROM planets WHERE num_of_moons > 1"
df_planets_more_than_1_moon = pd.read_sql_query(query_planets_more_than_1_moon, conn)
print("\nPlanets with More Than 1 Moon:")
print(df_planets_more_than_1_moon)

# 4. Select Planets Ordered by Mass (Descending)
query_planets_ordered_by_mass = "SELECT * FROM planets ORDER BY mass DESC"
df_planets_ordered_by_mass = pd.read_sql_query(query_planets_ordered_by_mass, conn)
print("\nPlanets Ordered by Mass (Descending):")
print(df_planets_ordered_by_mass)

# 5. Select Planets with Mass Greater Than 1
query_planets_mass_greater_than_1 = "SELECT * FROM planets WHERE mass > 1"
df_planets_mass_greater_than_1 = pd.read_sql_query(query_planets_mass_greater_than_1, conn)
print("\nPlanets with Mass Greater Than 1:")
print(df_planets_mass_greater_than_1)

# 6. Select Planets with Mass Between 0.5 and 5
query_planets_mass_between = "SELECT * FROM planets WHERE mass BETWEEN 0.5 AND 5"
df_planets_mass_between = pd.read_sql_query(query_planets_mass_between, conn)
print("\nPlanets with Mass Between 0.5 and 5:")
print(df_planets_mass_between)

# 7. Select Planets with Specific Colors (e.g., 'blue' or 'red')
query_planets_specific_colors = "SELECT * FROM planets WHERE color IN ('blue', 'red')"
df_planets_specific_colors = pd.read_sql_query(query_planets_specific_colors, conn)
print("\nPlanets with Specific Colors (blue, red):")
print(df_planets_specific_colors)

# 8. Count the Number of Planets
query_count_planets = "SELECT COUNT(*) AS num_planets FROM planets"
df_count_planets = pd.read_sql_query(query_count_planets, conn)
print("\nNumber of Planets:")
print(df_count_planets)

# 9. Average Number of Moons
query_avg_num_of_moons = "SELECT AVG(num_of_moons) AS avg_moons FROM planets"
df_avg_num_of_moons = pd.read_sql_query(query_avg_num_of_moons, conn)
print("\nAverage Number of Moons:")
print(df_avg_num_of_moons)

# 10. Total Mass of All Planets
query_total_mass = "SELECT SUM(mass) AS total_mass FROM planets"
df_total_mass = pd.read_sql_query(query_total_mass, conn)
print("\nTotal Mass of All Planets:")
print(df_total_mass)


All Planets:
   id     name       color  num_of_moons    mass  rings
0   1  Mercury        gray             0    0.55      0
1   2    Venus      yellow             0    0.82      0
2   3    Earth        blue             1    1.00      0
3   4     Mars         red             2    0.11      0
4   5  Jupiter      orange            68  317.90      0
5   6   Saturn       hazel            62   95.19      1
6   7   Uranus  light blue            27   14.54      1
7   8  Neptune   dark blue            14   17.15      1

Planets with Rings:
   id     name       color  num_of_moons   mass  rings
0   6   Saturn       hazel            62  95.19      1
1   7   Uranus  light blue            27  14.54      1
2   8  Neptune   dark blue            14  17.15      1

Planets with More Than 1 Moon:
   id     name       color  num_of_moons    mass  rings
0   4     Mars         red             2    0.11      0
1   5  Jupiter      orange            68  317.90      0
2   6   Saturn       hazel            62  

### 2. Select all columns for each planet whose `num_of_moons` is 0

In [3]:
# Your code here
# Write a SQL query to select all columns for planets with num_of_moons = 0
query_planets_no_moons = "SELECT * FROM planets WHERE num_of_moons = 0"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_no_moons = pd.read_sql_query(query_planets_no_moons, conn)

# Display the DataFrame
print(df_planets_no_moons)

   id     name   color  num_of_moons  mass  rings
0   1  Mercury    gray             0  0.55      0
1   2    Venus  yellow             0  0.82      0


### 3. Select the `name` and `mass` of each planet whose `name` has exactly 7 letters

In [4]:
# Your code here
# Write a SQL query to select the name and mass of planets with names of exactly 7 letters
query_planets_name_7_letters = "SELECT name, mass FROM planets WHERE LENGTH(name) = 7"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_name_7_letters = pd.read_sql_query(query_planets_name_7_letters, conn)

# Display the DataFrame
print(df_planets_name_7_letters)

      name    mass
0  Mercury    0.55
1  Jupiter  317.90
2  Neptune   17.15


### 4. Select all columns for each planet whose `mass` is greater than 1.00

In [5]:
# Your code here
# Write a SQL query to select all columns for planets with mass > 1.00
query_planets_mass_greater_than_1 = "SELECT * FROM planets WHERE mass > 1.00"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_mass_greater_than_1 = pd.read_sql_query(query_planets_mass_greater_than_1, conn)

# Display the DataFrame
print(df_planets_mass_greater_than_1)

   id     name       color  num_of_moons    mass  rings
0   5  Jupiter      orange            68  317.90      0
1   6   Saturn       hazel            62   95.19      1
2   7   Uranus  light blue            27   14.54      1
3   8  Neptune   dark blue            14   17.15      1


### 5. Select the `name` and `mass` of each planet whose `mass` is less than or equal to 1.00

In [6]:
# Your code here
# Write a SQL query to select the name and mass of planets with mass <= 1.00
query_planets_mass_less_equal_1 = "SELECT name, mass FROM planets WHERE mass <= 1.00"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_mass_less_equal_1 = pd.read_sql_query(query_planets_mass_less_equal_1, conn)

# Display the DataFrame
print(df_planets_mass_less_equal_1)

      name  mass
0  Mercury  0.55
1    Venus  0.82
2    Earth  1.00
3     Mars  0.11


### 6. Select the `name` and `mass` of each planet whose `mass` is between 0 and 50

In [7]:
# Your code here
# Write a SQL query to select the name and mass of planets with mass between 0 and 50
query_planets_mass_between_0_and_50 = "SELECT name, mass FROM planets WHERE mass BETWEEN 0 AND 50"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_mass_between_0_and_50 = pd.read_sql_query(query_planets_mass_between_0_and_50, conn)

# Display the DataFrame
print(df_planets_mass_between_0_and_50)

      name   mass
0  Mercury   0.55
1    Venus   0.82
2    Earth   1.00
3     Mars   0.11
4   Uranus  14.54
5  Neptune  17.15


### 7. Select all columns for planets that have at least one moon and a `mass` less than 1.00

***Hint:*** You can use `AND` to chain together two conditions in SQL, similar to `and` in Python

In [8]:
# Your code here
# Write a SQL query to select all columns for planets with at least one moon and mass < 1.00
query_planets_moons_and_mass = "SELECT * FROM planets WHERE num_of_moons > 0 AND mass < 1.00"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_moons_and_mass = pd.read_sql_query(query_planets_moons_and_mass, conn)

# Display the DataFrame
print(df_planets_moons_and_mass)

   id  name color  num_of_moons  mass  rings
0   4  Mars   red             2  0.11      0


### 8. Select the `name` and `color` of planets that have a `color` containing the string "blue"

In [9]:
# Your code here
# Write a SQL query to select the name and color of planets with color containing "blue"
query_planets_color_blue = "SELECT name, color FROM planets WHERE color LIKE '%blue%'"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_color_blue = pd.read_sql_query(query_planets_color_blue, conn)

# Display the DataFrame
print(df_planets_color_blue)


      name       color
0    Earth        blue
1   Uranus  light blue
2  Neptune   dark blue


### 9. Select the count of planets that don't have rings as `planets_without_rings`

Note: even though the schema states that `rings` is a `BOOLEAN` and the example table shows values `TRUE` and `FALSE`, SQLite does not actually support booleans natively. From the [documentation](https://www.sqlite.org/datatype3.html#boolean_datatype):

> SQLite does not have a separate Boolean storage class. Instead, Boolean values are stored as integers 0 (false) and 1 (true).

In [10]:
# Your code here
# Write a SQL query to count the planets without rings
query_planets_without_rings = "SELECT COUNT(*) AS planets_without_rings FROM planets WHERE rings = 0"

# Execute the query and fetch the data into a pandas DataFrame
df_planets_without_rings = pd.read_sql_query(query_planets_without_rings, conn)

# Display the DataFrame
print(df_planets_without_rings)

   planets_without_rings
0                      5


### 10. Select the name of all planets, along with a value `has_rings` that returns "Yes" if the planet does have rings, and "No" if it does not

In [11]:
# Your code here
# Write a SQL query to select the name of all planets and a has_rings indicator
query_planets_has_rings = """
SELECT name,
       CASE
           WHEN rings = 1 THEN 'Yes'
           ELSE 'No'
       END AS has_rings
FROM planets
"""

# Execute the query and fetch the data into a pandas DataFrame
df_planets_has_rings = pd.read_sql_query(query_planets_has_rings, conn)

# Display the DataFrame
print(df_planets_has_rings)

      name has_rings
0  Mercury        No
1    Venus        No
2    Earth        No
3     Mars        No
4  Jupiter        No
5   Saturn       Yes
6   Uranus       Yes
7  Neptune       Yes


## Summary

Congratulations! NASA is one step closer to embarking upon its mission to Mars. In this lab, You practiced writing `SELECT` statements that query a single table to get specific information. You also used other clauses and specified column names to cherry-pick the data we wanted to retrieve. 